types/wlr_virtual_pointer_v1.c

```diff
@@ -0,0 +1,327 @@
+	wl_resource_set_user_data(pointer->resource, NULL);
+	wlr_signal_emit_safe(&pointer->events.destroy, pointer);
+	wl_signal_init(&pointer->events.destroy);
+	wl_list_remove(&pointer->link);
+	free(pointer);
+}
+
+static const struct wlr_input_device_impl input_device_impl = {
+	.destroy = input_device_destroy
+};
+
+static const struct zwlr_virtual_pointer_v1_interface virtual_pointer_impl;
+
+static struct wlr_virtual_pointer_v1 *virtual_pointer_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwlr_virtual_pointer_v1_interface, &virtual_pointer_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void virtual_pointer_motion(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time,
+		wl_fixed_t dx, wl_fixed_t dy) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_motion event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.delta_x = wl_fixed_to_double(dx),
+		.delta_y = wl_fixed_to_double(dy),
+		.unaccel_dx = wl_fixed_to_double(dx),
+		.unaccel_dy = wl_fixed_to_double(dy),
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.motion, &event);
+}
+
+static void virtual_pointer_motion_absolute(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t x, uint32_t y,
+		uint32_t x_extent, uint32_t y_extent) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_motion_absolute event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.x = (double)x / x_extent,
+		.y = (double)y / y_extent,
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.motion_absolute, &event);
+}
+
+static void virtual_pointer_button(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t button,
+		uint32_t state) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_button event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.button = button,
+		.state = state ? WLR_BUTTON_PRESSED : WLR_BUTTON_RELEASED
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.button, &event);
+}
+
+static void virtual_pointer_axis(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t source,
+		wl_fixed_t value) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	pointer->axis_event.device = wlr_dev;
+	pointer->axis_event.time_msec = time;
+	pointer->axis_event.source = source;
```

Comment:
This logic doesn't handle this sequence of events properly:

```
axis(time=42, axis=vertical, value=69)
axis(time=42, axis=horizontal, value=69)
frame()
```

It should send two axis events, yet it overwrites the vertical event with the horizontal one. I'd suggest making `pointer->axis_event` an array with as many elements as there are axes (currently 2).